src/julia_to_gap.jl

```diff
@@ -93,148 +131,235 @@ function julia_to_gap(x::Rational{T}) where {T<:Integer}
+
+    rec = recursive && _needs_tracking_julia_to_gap(T)
+    if rec && recursion_dict === nothing
+        # Tracking of identical subobjects is needed,
+        # create the dictionary.
+        recursion_dict = RecDict()
+    end
+
+    # If we track identical subobjects then add `obj` to the dictionary.
+    if recursion_dict !== nothing
+      recursion_dict[obj] = ret_val
     end
+
+    # Set the subobjects.
     for i = 1:len
         x = obj[i]
         if x === nothing
             continue
         end
         if recursive
-            x = get!(recursion_dict, x) do
-                julia_to_gap(x, recursion_dict; recursive)
+            # Convert the subobjects.
+            # Do not care about findíng them in the dictionary
+            # or adding them to the dictionary,
+            # since their conversion method decides about that.
+            if rec
+              res = GapObj_internal(x, recursion_dict::RecDict, Val(recursive))
+            else
+              res = GapObj_internal(x, nothing, Val(recursive))
             end
+        else
+            res = x
         end
-        ret_val[i] = x
+        ret_val[i] = res
     end
+
+    return ret_val
+end
+
+## Sets
+function GapObj_internal(
+    obj::Set{T},
+    recursion_dict::GapCacheDict,
+    ::Val{recursive},
+) where {T, recursive}
+
+    recursive && recursion_dict !== nothing && haskey(recursion_dict, obj) && return recursion_dict[obj]
+
+    ret_val = GAP.NewPlist(length(obj))
+
+    rec = recursive && _needs_tracking_julia_to_gap(T)
+    if rec && recursion_dict === nothing
+        recursion_dict = RecDict()
+    end
+
+    if recursion_dict !== nothing
+        recursion_dict[obj] = ret_val
+    end
+
+    for x in obj
+        if recursive
+          if rec
+            res = GapObj_internal(x, recursion_dict::RecDict, Val(recursive))
+          else
+            res = GapObj_internal(x, nothing, Val(recursive))
+          end
+        else
+          res = x
+        end
+        Wrappers.Add(ret_val, res)
+    end
+    Wrappers.Sort(ret_val)
+    @assert Wrappers.IsSet(ret_val)
+
     return ret_val
 end
 
 ## Convert two dimensional arrays
-function julia_to_gap(
+function GapObj_internal(
     obj::Matrix{T},
-    recursion_dict::IdDict{Any,Any} = IdDict();
-    recursive::Bool = false,
-) where {T}
-    (rows, cols) = size(obj)
-    if haskey(recursion_dict, obj)
-        return recursion_dict[obj]
-    end
+    recursion_dict::GapCacheDict,
+    ::Val{recursive},
+) where {T, recursive}
+
+    recursive && recursion_dict !== nothing && haskey(recursion_dict, obj) && return recursion_dict[obj]
+
+    rows = size(obj, 1)
     ret_val = NewPlist(rows)
-    if recursive
-        recursion_dict[obj] = ret_val
+
+    rec = recursive && _needs_tracking_julia_to_gap(T)
+    if rec && recursion_dict === nothing
+        recursion_dict = RecDict()
     end
+
+    if recursion_dict !== nothing
+      recursion_dict[obj] = ret_val
+    end
+
```

Comment:
I wonder if we should do this:
```suggestion

    if !rec
      recursion_dict = nothing
    end

```
With that, I think we could replace the `if rec / else` in the for loop by the content of the `if` clause, i.e. just
```
    for i = 1:rows
        ret_val[i] = GapObj_internal(obj[i, :], recursion_dict::RecDict, Val(recursive))
    end
```

And then perhaps we could put this ugly but repetitive intro block into a macro or helper function?